### DISCUSSION TITLE
Step-by-step tutorial to hide internal edges of quadratic cells

### DISCUSSION THREAD

**Discussion by adtzlr** [pyvista/pyvista#5777] [Ideas]
### Hide internal edges of quadratic cells

Hi,

this is a manual way to hide the internal edges of (connected) quadratic cells (see also https://github.com/pyvista/pyvista/issues/867).

### Step-by-step explanation
1. Separate the cells of the mesh (see [`separate_cells()`](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.separate_cells.html))
2. Extract the surface with any nonlinear subdivision level >= 1 (see [`extract_surface(nonlinear_subdivision)`](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.extract_surface.html))
3. Extract the feature edges (see [`extract_feature_edges()`](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.extract_feature_edges.html))
4. Merge and plot both meshes

### Pseudocode or Screenshots

And here's the code. I used the resulting displacements of a finite element simulation on tri-quadratic hexahedrons.

```python
# !pip install felupe[all]
import felupe as fem

# finite element simulation: clamped uniaxial compression of a cube
cube = fem.Cube(n=4).add_midpoints_edges().add_midpoints_faces().add_midpoints_volumes()
region = fem.RegionTriQuadraticHexahedron(cube)
field = fem.FieldContainer([fem.Field(region, dim=3)])
boundaries, loadcase = fem.dof.uniaxial(field, clamped=True, move=-0.3)
solid = fem.SolidBody(umat=fem.NeoHooke(mu=1, bulk=5), field=field)
step = fem.Step(items=[solid], boundaries=boundaries)
job = fem.Job(steps=[step]).evaluate()

# a warped pyvista mesh
mesh = cube.view(point_data={"Displacement": field[0].values}).mesh
mesh = mesh.warp_by_vector("Displacement")
```

Thanks to @tkoyama010, this is an easy task in PyVista.

```python
import pyvista as pv

surface = mesh.separate_cells().extract_surface(nonlinear_subdivision=4)
edges = surface.extract_feature_edges()
(surface + edges).plot()
```

![image](https://github.com/pyvista/pyvista/assets/5793153/d08b7d28-b5cc-44d5-b36b-2b36419bb23f)

For black edges, add the two meshes to a `Plotter`.
```python
plotter = pv.Plotter()
plotter.add_mesh(surface)
plotter.add_mesh(edges, style="wireframe", color="black", line_width=3)
plotter.show()
```

![image](https://github.com/pyvista/pyvista/assets/5793153/1d0572ef-ac4e-4329-a678-3405195c7d5f)

<details><summary>For completeness, here is the initial code-block.</summary>

![image](https://github.com/pyvista/pyvista/assets/5793153/932c12f2-64f5-41da-8911-21550777c379)

```python
# first extract and subdivide the surface
view = field.view()
view.mesh = view.mesh.extract_surface(nonlinear_subdivision=4)

# plot the subdivided surface without edges
plotter = view.plot(show_edges=False)

# disconnect the mesh, create a new field and apply the point data
m = mesh.disconnect()
r = fem.RegionTriQuadraticHexahedron(m)
f = fem.Field(r, values=field[0].values[mesh.cells].reshape(-1, 3))

# extract the surface and the feature edges of the disconnected mesh
view = fem.FieldContainer([f]).view()
view.mesh = view.mesh.extract_surface(nonlinear_subdivision=4).extract_feature_edges()

# plot the feature edges
plotter = view.plot(plotter=plotter, style="wireframe", color="black", line_width=3)
plotter.show()
```

</details>

**Comment by tkoyama010** [comment]
Awesome tutorial!

> Sorry, I don't know how to perform a `disconnect()`-operation directly in pyvista. I'm sure this is easy to do. Meanwhile, let's use FElupe.

[separate_cells](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.separate_cells.html#pyvista.DataSetFilters.separate_cells) is what you are looking for?

**Reply by adtzlr** [reply]
Oh yes, perfect! I'll update the code-snippet accordingly. Now FElupe only provides a warped mesh, everything else is done in PyVista. Hopefully this is helpful.

@tkoyama010 @banesullivan I think this would be a really helpful tutorial or example for the documentation in combination with the tri-quadratic hexahedron example.

```python
from pyvista import examples
mesh = examples.download_tri_quadratic_hexahedron()
```

**Reply by tkoyama010** [reply]
The following PyVista tutorial folder introduces external examples of PyVista. Would you like to be introduced to FElupe?
https://github.com/pyvista/pyvista-tutorial/tree/main/tutorial/05_action

**Reply by akaszynski** [reply]
Note that you can reduce the visualization artifact of edges being obscured by the underlying mesh by using `SetResolveCoincidentTopologyToPolygonOffset`:

![tmp](https://github.com/pyvista/pyvista/assets/11981631/d2ffac86-193b-41dc-8891-599ff5295f6b)


```py
import pyvista as pv

from pyvista import examples
mesh = examples.download_tri_quadratic_hexahedron()
surface_sep = mesh.separate_cells().extract_surface(nonlinear_subdivision=4)
edges = surface_sep.extract_feature_edges()

surface = mesh.extract_surface(nonlinear_subdivision=4)

pl = pv.Plotter()
pl.add_mesh(surface, smooth_shading=True, split_sharp_edges=True)
actor = pl.add_mesh(edges, color='k', line_width=3)
actor.mapper.SetResolveCoincidentTopologyToPolygonOffset()
pl.show()
```

**Reply by tkoyama010** [reply]
Thanks, you are the best! I added #5781 inspired your code. Now, we can use the following property.
```python
actor.mapper.resolve = "polygon_offset"
```

**Reply by adtzlr** [reply]
Thank you both @tkoyama010 and @akaszynski for pointing out the details! The result is amazing 💯

**Reply by adtzlr** [reply]
@akaszynski Is `pl.ren_win` equal to [`pl.render_window`](https://docs.pyvista.org/version/stable/api/plotting/_autosummary/pyvista.Plotter.render_window.html)? Is this required here?

**Reply by akaszynski** [reply]
Nope, fixed. Thanks for pointing that out.

**Comment by tkoyama010** [comment]
As a FEA person, I find FElupe's page excellent. @akaszynski, you might also be interested in it.
[![image](https://github.com/pyvista/pyvista/assets/7513610/8114e5ce-d85c-4e0c-8da9-bbf0b61ff114)](https://felupe.readthedocs.io/en/latest/examples/index.html)

**Reply by akaszynski** [reply]
Docs look great! It's also nice to see the interactive docs rendered.

**Reply by adtzlr** [reply]
Thanks @tkoyama010 and @akaszynski 🎉 , but there is still a lot of room for improvement 📝  (as always...)
